src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+                }
+                catch (Exception ex)
+                {
+                    // Report thrown exceptions
+                    Utilities.WriteLine($"The fix '{fix.CodeFixEquivalenceKey}' threw an exception after {stopwatch.ElapsedMilliseconds}ms:", ConsoleColor.Yellow);
+                    Utilities.WriteLine(ex.ToString(), ConsoleColor.Yellow);
+                }
+            }
+        }
+
+        private static async Task TestCodeFixesAsync(Stopwatch stopwatch, Solution solution, IEnumerable<Assembly> analyzerAssemblies, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            Console.WriteLine("Calculating fixes");
+
+            List<CodeAction> fixes = new List<CodeAction>();
+
+            var codeFixers = GetAllCodeFixers(analyzerAssemblies);
+
+            foreach (var item in diagnostics)
+            {
+                foreach (var codeFixer in codeFixers.GetValueOrDefault(item.Id, ImmutableList.Create<CodeFixProvider>()))
+                {
+                    fixes.AddRange(await GetFixesAsync(solution, codeFixer, item, cancellationToken).ConfigureAwait(false));
+                }
+            }
+
+            Console.WriteLine($"Found {fixes.Count} potential code fixes");
+
+            Console.WriteLine("Calculating changes");
+
+            stopwatch.Restart();
+
+            object lockObject = new object();
+
+            Parallel.ForEach(fixes, fix =>
+            {
+                try
+                {
+                    fix.GetOperationsAsync(cancellationToken).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
+                {
+                    // Report thrown exceptions
+                    lock (lockObject)
+                    {
+                        Utilities.WriteLine($"The fix '{fix.Title}' threw an exception:", ConsoleColor.Yellow);
+                        Utilities.WriteLine(ex.ToString(), ConsoleColor.Red);
+                    }
+                }
+            });
+
+            Console.WriteLine($"Calculating changes completed in {stopwatch.ElapsedMilliseconds}ms");
+        }
+
+        private static async Task<IEnumerable<CodeAction>> GetFixesAsync(Solution solution, CodeFixProvider codeFixProvider, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            List<CodeAction> codeActions = new List<CodeAction>();
+
+            await codeFixProvider.RegisterCodeFixesAsync(new CodeFixContext(solution.GetDocument(diagnostic.Location.SourceTree), diagnostic, (a, d) => codeActions.Add(a), cancellationToken)).ConfigureAwait(false);
+
+            return codeActions;
+        }
+
+        private static Task<Statistic> GetAnalyzerStatisticsAsync(IEnumerable<Project> projects, CancellationToken cancellationToken)
```

Comment:
if it is going to use GetAwaiter and GetResult, this method doesn't need to be Async method. or this code can be rewritten to Task.WhenAll(tasks)

Reply:
I think this method name is little bit misleading. this is not actually AnalyzerStatistics, but Solution Statistics.

Analyzer statistics will be this (http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerTelemetry.cs,10) you can get from CompilationWithAnalyzer.

and that info is probably good enough to have options for. since we can see what analyzer API the analyzer is using.

Reply:
> if it is going to use GetAwaiter and GetResult, this method doesn't need to be Async method. or this code can be rewritten to Task.WhenAll(tasks)

:arrow_right: Task.WhenAll will not force the parallelization of tasks which do not yield. `Parallel.ForEach` will.

Reply:
hmm.. get GetAwaiter().GetResult() is used since the inner method is actually Async method which call await SyntaxTree and etc. which is real Async method. and that means, blocking parallelization can be worse than await since, that require current thread to be blocked, it require other thread to run parsing tree and etc, and then when they are done come back to this blocked thread. so, you can easily get into thread starvation situation which affect perf negatively.

I think it is better to use await for real Async method. adding ConfigureAwait(false) I don't believe solve thread starvation issue, since it will still waste 1 thread for blocking. it only let it to resume from different thread then current one.

...

anyway, regardless, if it is going to use Parallel.For, the method should be sync method, rather than Async which doesn't have any Async in it and return Task.From(...)